### PR TITLE
Article store events

### PR DESCRIPTION
**Pull request by seanwiseman**
Implements requirements from [#31](https://github.com/libero/walking-skeleton/issues/31).

### CODE REVIEW

**Comment by seanwiseman** [comment]
Currently sends messages for the following events:

`article.delete.[started|completed|failed]`
`article.version.[delete|post|put].[started|completed|failed]`

You should see them populate the dashboard.

File: dashboard/config/packages/old_sound_rabbit_mq.yaml [hunk 1]
```diff
@@ -8,6 +8,6 @@ old_sound_rabbit_mq:
             enable_logger: true
             exchange_options:
                 name: 'dashboard'
-                type: direct
+                type: fanout
             queue_options:
-                name: 'dashboard'
+                name: 'articles'
```
Comment (thewilkybarkid) [inline dashboard/config/packages/old_sound_rabbit_mq.yaml]: Should be the exchange's name changed?

Comment (seanwiseman) [reply dashboard/config/packages/old_sound_rabbit_mq.yaml]: As we are not producing any messages (from what I can see) from the dashboard at present, I am not sure of the exchange's purpose here 🤔 .  Maybe was just for testing.

You only need the queue name (`articles` in this case) to receive the messages. The `exchange_options` entry can be removed if we are not producing messages from the dashboard.

Comment (thewilkybarkid) [reply dashboard/config/packages/old_sound_rabbit_mq.yaml]: Should be the exchange that the queue is bound to (this is under the `consumers` config rather than `producers`).

File: article-store/events/conf.py [hunk 2]
```diff
@@ -0,0 +1,21 @@
...
+import pika
+
+
+RABBITMQ_HOST = os.environ.get('RABBITMQ_HOST', 'rabbitmq')  # event-bus
+RABBITMQ_PORT = os.environ.get('RABBITMQ_PORT', 5672)
+RABBITMQ_PASSWORD = os.environ.get('RABBITMQ_PASSWORD')
+RABBITMQ_USER = os.environ.get('RABBITMQ_USER')
+
+BROKER_CREDENTIALS = pika.PlainCredentials(RABBITMQ_USER, RABBITMQ_PASSWORD)
+BROKER_PARAMS = pika.ConnectionParameters(host=RABBITMQ_HOST, credentials=BROKER_CREDENTIALS)
```
Comment (thewilkybarkid) [inline article-store/events/conf.py]: Same as https://github.com/libero/walking-skeleton/pull/25#discussion_r189183325

File: article-store/events/utils.py [hunk 3]
```diff
@@ -0,0 +1,122 @@
...
+    try:
+        connection = pika.BlockingConnection(parameters=BROKER_PARAMS)
+        yield connection.channel()
+        connection.close()
+    except ConnectionClosed as err:
+        print(err)
+        raise
+
+
+def setup_exchanges_and_queues(func) -> Callable[..., None]:
```
Comment (thewilkybarkid) [inline article-store/events/utils.py]: We don't need a queue here? Though guess this is just copied from earlier work?

File: dashboard/config/packages/old_sound_rabbit_mq.yaml (hunk 1 above)
Comment (thewilkybarkid) [reply dashboard/config/packages/old_sound_rabbit_mq.yaml]: Can see this is actually working... taking a proper look.

File: article-store/events/utils.py (hunk 3 above)
Comment (seanwiseman) [reply article-store/events/utils.py]: Yes, it was brought in from previous work, but also makes sure the queue(s) you are binding the exchange(s) to exists before the actual binding. This may be subject to change if we can guarantee a queues existence but for now, I think we have to do it this way.

Comment (thewilkybarkid) [reply article-store/events/utils.py]: But in this case we don't need the `articles` queue?

File: article-store/events/conf.py (hunk 2 above)
Comment (seanwiseman) [reply article-store/events/conf.py]: Done 👍

File: article-store/events/utils.py (hunk 3 above)
Comment (seanwiseman) [reply article-store/events/utils.py]: No, it only actually needs an exchange to publish.

**Comment by seanwiseman** [comment]
Are we going to want the `article.version.published` event available in this PR?

File: article-store/Pipfile [hunk 4]
```diff
@@ -10,12 +10,14 @@ name = "pypi"
 django = "~=2.0.5"
 django-debug-toolbar = "~=1.9.1"
 djangorestframework = "~=3.8.2"
-psycopg2-binary = "*"
+"psycopg2-binary" = "*"
```
Comment (giorgiosironi) [inline article-store/Pipfile:13]: pipenv strikes back - not a problem

File: article-store/articles/tests/test_api/test_article_api.py [hunk 5]
```diff
@@ -227,7 +235,7 @@ def test_can_get_article_content_without_giving_part_name(self,
                                                               content_es_front: Content,
                                                               article_0065_es_front_xml: str):
 
-        url = f'/articles/{article.id}/latest'
+        url = f'/articles/{article.id}/versions/latest'
```
Comment (giorgiosironi) [inline article-store/articles/tests/test_api/test_article_api.py:238]: changes like these should usually be a separate PR?

Comment (giorgiosironi) [reply article-store/articles/tests/test_api/test_article_api.py:238]: not that the history of the walking skeleton itself should be that clean in the long run, projects are going to be extracted in other repositories

File: article-store/events/conf.py [hunk 6]
```diff
@@ -0,0 +1,17 @@
+import os
+
+import pika
+
+
+RABBITMQ_URL = os.environ.get('RABBITMQ_URL')
+
+BROKER_PARAMS = pika.connection.URLParameters(RABBITMQ_URL)
+
+ARTICLES_QUEUE = os.environ.get('ARTICLES_QUEUE', 'articles')
```
Comment (giorgiosironi) [inline article-store/events/conf.py]: I agree with other comments that a queue should not be necessary here. If we want to set up one for testing, it should be clearly named as `articles-dummy` or similar

File: article-store/events/utils.py [hunk 7]
```diff
@@ -0,0 +1,122 @@
...
+    :param article_id: str
+    :param article_version: int
+    :return:
+    """
+    with get_channel() as channel:
+        message = create_message(msg_type=msg_type, identifier=article_id,
+                                 data={'article_version': article_version})
+
+        channel.basic_publish(exchange=ARTICLES_EXCHANGE,
+                              routing_key="",
```
Comment (giorgiosironi) [inline article-store/events/utils.py:92]: didn't understand this can be empty, good to know

File: article-store/articles/tests/test_api/test_article_api.py (hunk 5 above)
Comment (seanwiseman) [reply article-store/articles/tests/test_api/test_article_api.py:238]: Yes I agree.

File: article-store/events/utils.py [hunk 8]
```diff
@@ -0,0 +1,122 @@
...
+
+def create_message(msg_type: str, identifier: str, data: Dict) -> Dict:
+    """Create a message `dict` based on a standard schema.
+
+    :return: Dict
+    """
+    return {
+        "eventId": str(uuid.uuid1()),
+        "happenedAt": strftime("%Y-%m-%dT%H:%M:%S+00:00", gmtime()),
+        "aggregate": {
```
Comment (giorgiosironi) [inline article-store/events/utils.py:35]: may be changed to `entity` in the schema (not yet)

File: article-store/events/utils.py [hunk 9]
```diff
@@ -0,0 +1,122 @@
...
+    :return:
+    """
+    @wraps(func)
+    def wrapper(*args, **kwargs):
+        with get_channel() as channel:
+            for queue_name in DEFAULT_QUEUES:
+                channel.queue_declare(queue=queue_name, durable=True)
+
+                for exchange in DEFAULT_QUEUES[queue_name]:
+                    channel.exchange_declare(exchange=exchange, exchange_type='fanout', durable=True)
```
Comment (giorgiosironi) [inline article-store/events/utils.py]: exchanges (only `ARTICLES_EXCHANGE`) should be declared even if there are no queues to be bound to them; this code can be pulled up to the start of the `with get_channel()` block

File: article-store/events/utils.py [hunk 10]
```diff
@@ -0,0 +1,122 @@
...
+def message_publisher(msg_type: str, article_id: str, article_version: int = None) -> ContextManager[None]:
+    """Wrapper to send started/completed/failed messages for a given article `msg_type`.
+
+    :param msg_type: str
+    :param article_id: str
+    :param article_version: int
+    :return:
+    """
+
+    send_message = partial(send_article_message, article_id=article_id, article_version=article_version)
```
Comment (giorgiosironi) [inline article-store/events/utils.py:107]: this is neat

Comment (giorgiosironi) [reply article-store/events/utils.py:107]: (this context manager in general)

File: article-store/events/utils.py [hunk 11]
```diff
@@ -0,0 +1,122 @@
...
+    """Create a message `dict` based on a standard schema.
+
+    :return: Dict
+    """
+    return {
+        "eventId": str(uuid.uuid1()),
+        "happenedAt": strftime("%Y-%m-%dT%H:%M:%S+00:00", gmtime()),
+        "aggregate": {
+            "service": "article-store",
+            "name": "article-version",
```
Comment (giorgiosironi) [inline article-store/events/utils.py:37]: may be `article` if this is an `article.delete`-related event?

Comment (seanwiseman) [reply article-store/events/utils.py:37]: mmm, the `article.delete` or `article.version.delete` is currently defined in the `type`, does `aggregate.name` have to change with it as well? or could `aggregate.name` just always be `article`?

File: dashboard/config/packages/old_sound_rabbit_mq.yaml (hunk 1 above)
Comment (thewilkybarkid) [reply dashboard/config/packages/old_sound_rabbit_mq.yaml]: I had worked out what was happening, must not have submitted my follow-up. 🤔

This results in an exchange called 'dashboard' and a bound queue called 'articles', the article store _was_ creating an exchange called 'articles' and a bound queue called 'articles'. This works as the queues matched, the 'dashboad' exchange was just superfluous.

Now the article store doesn't create a queue, this probably no longer works. `exchange_options.name` needs to be  'articles' and `queue_options.name` needs to be 'dashboard'. That binds the queue for this app to the exchange provided by the article store.

File: article-store/events/utils.py (hunk 10 above)
Comment (thewilkybarkid) [reply article-store/events/utils.py:107]: Agreed!

File: article-store/events/utils.py [hunk 12]
```diff
@@ -0,0 +1,118 @@
...
+    """
+
+    send_message = partial(send_article_message, article_id=article_id, article_version=article_version)
+
+    try:
+        send_message(f'{msg_type}.started')
+        yield
+        send_message(f'{msg_type}.completed')
+    except Exception:
+        send_message(f'{msg_type}.failed')
```
Comment (thewilkybarkid) [inline article-store/events/utils.py:114]: Shouldn't this `raise`?

File: article-store/events/conf.py [hunk 13]
```diff
@@ -3,15 +3,12 @@
 import pika
 
 
-RABBITMQ_URL = os.environ.get('RABBITMQ_URL')
+RABBITMQ_URL = os.environ.get('RABBITMQ_URL', '')
 
-BROKER_PARAMS = pika.connection.URLParameters(RABBITMQ_URL, '')
+BROKER_PARAMS = pika.connection.URLParameters(RABBITMQ_URL)
 
 ARTICLES_DUMMY_QUEUE = os.environ.get('ARTICLES_DUMMY_QUEUE', 'articles-dummy')
```
Comment (thewilkybarkid) [inline article-store/events/conf.py]: Can be removed?

File: article-store/events/utils.py (hunk 12 above)
Comment (seanwiseman) [reply article-store/events/utils.py:114]: Good spot.

File: dashboard/config/packages/old_sound_rabbit_mq.yaml (hunk 1 above)
Comment (seanwiseman) [reply dashboard/config/packages/old_sound_rabbit_mq.yaml]: Ok, I have made the change and retested and it works fine.

File: article-store/Pipfile (hunk 4 above)
Comment (thewilkybarkid) [reply article-store/Pipfile:13]: https://github.com/pypa/pipenv/issues/1933

File: article-store/events/utils.py (hunk 11 above)
Comment (giorgiosironi) [reply article-store/events/utils.py:37]: I'd say comment out the `aggregate` field for now and wait for https://github.com/libero/walking-skeleton/pull/28/files to add it later when stable

Comment (seanwiseman) [reply article-store/events/utils.py:37]: Ok, the article id is in `aggregate.identifier` currently. Should it be moved elsewhere for the time being?

Comment (giorgiosironi) [reply article-store/events/utils.py:37]: No need to do big changes, just adding a comment here about this field being experimental and about to evolve/being replaced would be fine.

File: dashboard/config/packages/old_sound_rabbit_mq.yaml [hunk 14]
```diff
@@ -7,7 +7,7 @@ old_sound_rabbit_mq:
             callback: Libero\Dashboard\Consumer\EventsConsumer
             enable_logger: true
             exchange_options:
-                name: 'dashboard'
-                type: direct
+                name: 'articles'
+                type: fanout
```
Comment (giorgiosironi) [inline dashboard/config/packages/old_sound_rabbit_mq.yaml:11]: not sure why the dashboard should know about the type of the exchange, is it required to subscribe to it? The risk is a race condition where the dashboard preempts the article store in creating the exchange, which shouldn't belong to that service.

Comment (thewilkybarkid) [reply dashboard/config/packages/old_sound_rabbit_mq.yaml:11]: Untested, but looks like you can set `declare: false` and it won't create it.

Comment (seanwiseman) [reply dashboard/config/packages/old_sound_rabbit_mq.yaml:11]: Trying to test this now. Which level is `declare` configured at in the file?

All combinations I have tried so far have still lead to `You must provide an exchange type  
`.

Comment (thewilkybarkid) [reply dashboard/config/packages/old_sound_rabbit_mq.yaml:11]: Same level (https://github.com/php-amqplib/RabbitMqBundle/blob/cf67adaa4e306d8e9cb6855a72d79263b425ded8/DependencyInjection/Configuration.php#L371-L389), so should be:

```yaml
exchange_options:
    name: 'articles'
    type: fanout
    declare: false
```

Comment (seanwiseman) [reply dashboard/config/packages/old_sound_rabbit_mq.yaml:11]: Ah ok, thank you. I thought the `declare` meant we did not have to specify a value for the `type`.

OK, this works. It does not create the exchange when `declare: false` is present.

Comment (thewilkybarkid) [reply dashboard/config/packages/old_sound_rabbit_mq.yaml:11]: I was surprised too, feels redundant.

Comment (seanwiseman) [reply dashboard/config/packages/old_sound_rabbit_mq.yaml:11]: This however causes the following condition: (after a clean startup with no previous volume data)

- `dasboard` queue present
- `articles` exchange not present until the store creates it
- `dashboard` therefore cannot bind to it: (error on startup)
```
In AMQPChannel.php line 188:

NOT_FOUND - no exchange 'articles' in vhost '/'  
```
- There does not seem to be another attempt to bind (though are we expecting it to?)
- Even after a manual binding (via the rabbitMQ UI) of `dashboard` --> `articles` and new events being sent, the dashboard (app) still does not consume the events. I can however see the events pending on the queue (via the rabbitMQ UI) so they are being delivered to the `dashboard` queue.

Will probably benefit from an in person discussion.

Comment (giorgiosironi) [reply dashboard/config/packages/old_sound_rabbit_mq.yaml:11]: Yes, we've seen this before https://github.com/libero/walking-skeleton/pull/25/files#diff-bca2c15ea701ffde5b33224920caa539R39

Comment (seanwiseman) [reply dashboard/config/packages/old_sound_rabbit_mq.yaml:11]: As discussed, I will not implement the `declare: false`, it will be handled in another pr along with a method of checking the exchange is declared before trying to bind.

**Comment by seanwiseman** [comment]
:shipit: ?

**Review by giorgiosironi — APPROVED** [review]
Event data needs to be removed/refined, but let's do it separately not to hold this

**Review by thewilkybarkid — APPROVED** [review]
🎉
